### PR TITLE
opt: fix missing filters after join reordering

### LINKED ISSUE #76522: opt: join reordering can produce incorrect query plans
Join reordering can produce query plans where filters in the original plan are missing, causing incorrect query results.

Below is a reproduction of the issue. Notice that the `t2.should_not_be_eliminated = t5.should_not_be_eliminated` filter is missing from the final query plan.

```
exec-ddl
CREATE TABLE t1 (
	a INT NOT NULL,
	b INT NOT NULL,
	PRIMARY KEY (a ASC, b ASC)
)
----

exec-ddl
CREATE TABLE t2 (
	a INT NOT NULL,
	c INT,
	should_not_be_eliminated INT,
	PRIMARY KEY (a ASC)
);
----

exec-ddl
CREATE TABLE t3 (
	a INT NOT NULL,
	d INT NOT NULL,
	PRIMARY KEY (a ASC, d ASC)
);
----

exec-ddl
CREATE TABLE t4 (
	e INT NOT NULL,
	f INT,
	g INT,
	PRIMARY KEY (e ASC)
);
----

exec-ddl
CREATE TABLE t5 (
	h INT NOT NULL,
	f INT NOT NULL,
	g INT NOT NULL,
	b INT,
	should_not_be_eliminated INT,
	c INT,
	PRIMARY KEY (h ASC, f ASC, g ASC)
);
----

# Give t1 many rows where a has many distincts.
exec-ddl
ALTER TABLE t1 INJECT STATISTICS '[
    {
        "columns": [
            "a"
        ],
        "created_at": "2022-01-17 12:51:38.433911",
        "distinct_count": 9161427,
        "null_count": 0,
        "row_count": 44484238
    }
]';
----

# Give t2 many rows where a has many distincts.
exec-ddl
ALTER TABLE t2 INJECT STATISTICS '[
    {
        "columns": [
            "a"
        ],
        "created_at": "2022-01-17 12:51:38.433911",
        "distinct_count": 17014025,
        "null_count": 0,
        "row_count": 17024553
    }
]';
----

# Give t3 many rows where a has many distincts.
exec-ddl
ALTER TABLE t3 INJECT STATISTICS '[
    {
        "columns": [
            "a"
        ],
        "created_at": "2022-01-17 12:51:38.433911",
        "distinct_count": 17187349,
        "null_count": 0,
        "row_count": 18138540
    }
]';
----

# Give t4 many rows where e has many distincts.
exec-ddl
ALTER TABLE t4 INJECT STATISTICS '[
    {
        "columns": [
            "e"
        ],
        "created_at": "2022-01-17 12:51:38.433911",
        "distinct_count": 346919,
        "null_count": 0,
        "row_count": 346109
    }
]';
----

# Give t5 few rows.
exec-ddl
ALTER TABLE t5 INJECT STATISTICS '[
    {
        "columns": [
            "h"
        ],
        "created_at": "2022-01-17 12:51:38.433911",
        "distinct_count": 119,
        "null_count": 0,
        "row_count": 119
    }
]';
----

# Notice that some of the filters are missing entirely:
#
#   t5.c = t2.c
#   t2.should_not_be_eliminated = t5.should_not_be_eliminated
#
opt
SELECT
  t2.a
FROM
  t1
  INNER JOIN t2 ON t1.a = t2.a
  INNER JOIN t3 ON t1.a = t3.a
  INNER JOIN t4 ON t3.d = t4.e
  INNER JOIN t5 ON
      t4.f = t5.f
      AND t4.g = t5.g
      AND t5.b = t1.b
      AND t5.c = t2.c
WHERE
  t1.a = 123456 AND t2.should_not_be_eliminated = t5.should_not_be_eliminated;
----
project
 ├── columns: a:5!null
 └── inner-join (lookup t1)
      ├── columns: t1.a:1!null t1.b:2!null t2.a:5!null t2.c:6!null t2.should_not_be_eliminated:7!null t3.a:10!null d:11!null e:14!null t4.f:15!null t4.g:16!null t5.f:20!null t5.g:21!null t5.b:22!null t5.should_not_be_eliminated:23!null t5.c:24!null
      ├── key columns: [5 22] = [1 2]
      ├── lookup columns are key
      ├── inner-join (lookup t2)
      │    ├── columns: t2.a:5!null t2.c:6 t2.should_not_be_eliminated:7 t3.a:10!null d:11!null e:14!null t4.f:15!null t4.g:16!null t5.f:20!null t5.g:21!null t5.b:22 t5.should_not_be_eliminated:23 t5.c:24
      │    ├── key columns: [10] = [5]
      │    ├── lookup columns are key
      │    ├── inner-join (hash)
      │    │    ├── columns: t3.a:10!null d:11!null e:14!null t4.f:15!null t4.g:16!null t5.f:20!null t5.g:21!null t5.b:22 t5.should_not_be_eliminated:23 t5.c:24
      │    │    ├── scan t5
      │    │    │    └── columns: t5.f:20!null t5.g:21!null t5.b:22 t5.should_not_be_eliminated:23 t5.c:24
      │    │    ├── inner-join (lookup t4)
      │    │    │    ├── columns: t3.a:10!null d:11!null e:14!null t4.f:15 t4.g:16
      │    │    │    ├── key columns: [11] = [14]
      │    │    │    ├── lookup columns are key
      │    │    │    ├── scan t3
      │    │    │    │    ├── columns: t3.a:10!null d:11!null
      │    │    │    │    └── constraint: /10/11: [/123456 - /123456]
      │    │    │    └── filters (true)
      │    │    └── filters
      │    │         ├── t4.f:15 = t5.f:20
      │    │         └── t4.g:16 = t5.g:21
      │    └── filters
      │         └── t2.a:5 = 123456
      └── filters
           └── t1.a:1 = 123456
```

### PR DESCRIPTION
**Pull request by mgartner**
#### opt: add TES, SES, and rules to reorderjoins

This commit updates the output of the `reorderjoins` opt test command to
display the initial state of the `JoinOrderBuilder`. It adds additional
information to the output including the TES, SES, and conflict rules for
each edge.

Release note: None

#### opt: fix missing filters after join reordering

This commit eliminates logic in the `assoc`, `leftAsscom`, and
`rightAsscom` functions in the join order builder that aimed to prevent
generating "orphaned" predicates, where one or more referenced relations
are not in a join's input. In rare cases, this logic had the side effect
of creating invalid conflict rules for edges, which could prevent valid
predicates from being added to reordered join trees.

It is safe to remove these conditionals because they are unnecessary.
The CD-C algorithm already prevents generation of orphaned predicates by
checking that the total eligibility set (TES) is a subset of a join's
input vertices. In our implementation, this is handled by the
`checkNonInnerJoin` and `checkInnerJoin` functions.

Fixes #76522

Release note (bug fix): A bug has been fixed which caused the query optimizer
to omit join filters in rare cases when reordering joins, which could
result in incorrect query results. This bug was present since v20.2.

### CODE REVIEW

**Comment by mgartner** [comment]
@DrewKimball We've run into a bug where the join reorderer incorrectly eliminates join filters. Here's a test case that shows the problem: https://gist.github.com/mgartner/ff9857369f61d4d7bb7458d536c57664

Do the changes proposed here make sense? The logic behind them is: if we are reordering inner joins, the vertex sets of the original joins shouldn't whether or not the filters in two edges are associative.

This seems to solve the bug, without breaking any test cases. I'm curious if you remember any specific example that required theses lines of code.

cc @andy-kimball since he worked closely with some of this logic too.

**Comment by DrewKimball** [comment]
I can take a closer look at this in a bit, but for now I don't think those checks should be removed. The `checkProperty` function only checks `assoc`, `leftAsscom` etc. based on the join types, not the columns referenced by the filters. We still have to avoid pushing filters down below the columns they filter on. I'm suspecting the code that prevents adding redundant filters to the join: https://github.com/cockroachdb/cockroach/blob/c7446c5918673c4503f0260d74a708696482e1d1/pkg/sql/opt/xform/join_order_builder.go#L496-L499

**Comment by mgartner** [comment]
Thanks for taking a look. We really appreciate your help here!

> We still have to avoid pushing filters down below the columns they filter on.

Doesn't this check occur here: https://github.com/cockroachdb/cockroach/blob/7e83c054206242d1a50b25869238163cbe7328fb/pkg/sql/opt/xform/join_order_builder.go#L495

---

> I'm suspecting the code that prevents adding redundant filters to the join:

That's what I thought originally, but it doesn't appear to be the issue. We think the cause might be incorrect conflict rules that have been added to edges. Rebecca modified the `reorderjoins` test output to include the rules. The rules here seem incorrect: https://gist.github.com/mgartner/9401df94ca26124ac6b5d915da026632#file-join_reorder-txt-L273-L275

Consider`t5.b = t1.b [D -> C] [inner]`. This filter can be evaluated with just `A` and `E`, so the `D -> C` rule doesn’t seem valid.

This PR is an attempt to eliminate these allegedly bad conflict rules.

**Comment by DrewKimball** [comment]
>Doesn't this check occur here

I guess it does, since the referenced columns are encoded in the TES. We shouldn't be adding conflict rules based on which columns are referenced, since that information is already in the SES and TES. So I think this change is correct. Nice job tracking that down!

**Review by DrewKimball — COMMENTED** [review]
:lgtm:

(once that test case is added)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-:-MvW0-P-AJPJE41Nfm8z:b-6emhas)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Comment by mgartner** [comment]
This is ready for another look. I added a regression test for the bug, and a unit test that previously failed because an unnecessary conflict rule was created. I also added a commit that adds the SES, TES, and conflict rules of each edge to the `reorderjoins` test output.

**Review by rytaft — COMMENTED** [review]
Nice!

> Reviewed 4 of 4 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-:-Mvu8WMWAnkMiQlcY6Mu:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 887 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-Mvu9dCQ72CeCLXDt5iK:-Mvu9dCQ72CeCLXDt5iL:b5ou5pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa21d7472b641e7c7b688ede64e1b9bbe6af7628/pkg/sql/opt/xform/testdata/rules/join_order#L887)):*
> ```
>   B: scan cy
>   C: select
>    ├── scan dz
> ```

nit: looks like the formatting is a bit weird here
___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 2086 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-Mvu9rLAEYduvcZYOGAm:-Mvu9rLAEYduvcZYOGAn:b-6v9gq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa21d7472b641e7c7b688ede64e1b9bbe6af7628/pkg/sql/opt/xform/testdata/rules/join_order#L2086)):*
> ```
> Edges
>   a2.b = a3.b [inner, ses=CD, tes=CD, rules=()]
>   a3.a = a4.a [inner, ses=DE, tes=DE, rules=()]
> ```

It's a bit concerning that it seems we don't have any tests here with non-empty rules. Can you try to add one?
___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 2140 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-MvuAfmC1kiOVswpFKEe:-MvuAfmC1kiOVswpFKEf:b-vqgz40) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffac034236245112d50c59d03a4a4e9bdff87677/pkg/sql/opt/xform/testdata/rules/join_order#L2140)):*
> ```
> 	a INT NOT NULL,
> 	b INT NOT NULL,
> 	PRIMARY KEY (a ASC, b ASC)
> ```

nit: remove tabs

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 887 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-Mvu9dCQ72CeCLXDt5iK:-MvuFMqk-6nL8NuS8jDM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa21d7472b641e7c7b688ede64e1b9bbe6af7628/pkg/sql/opt/xform/testdata/rules/join_order#L887)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: looks like the formatting is a bit weird here
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 2086 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-Mvu9rLAEYduvcZYOGAm:-MvuPbEc7qdiEf71UMIy:b-s5ddo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa21d7472b641e7c7b688ede64e1b9bbe6af7628/pkg/sql/opt/xform/testdata/rules/join_order#L2086)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

It's a bit concerning that it seems we don't have any tests here with non-empty rules. Can you try to add one?
</blockquote></details>

I've added one test with a conflict rule on an edge. Conflict rules are somewhat uncommon because of some optimizations to reduce them into the TES, which is mentioned in section 5.5 of [the paper](https://dl.acm.org/doi/10.1145/2463676.2465314): https://github.com/cockroachdb/cockroach/blob/24d886cd713583fd6a4154e48bdbacad9eeb31f9/pkg/sql/opt/xform/join_order_builder.go#L1191-L1201
___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 2140 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-MvuAfmC1kiOVswpFKEe:-MvuDpLj-pHS2DxXhrE2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffac034236245112d50c59d03a4a4e9bdff87677/pkg/sql/opt/xform/testdata/rules/join_order#L2140)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: remove tabs
</blockquote></details>

Done.

**Review by msirek — APPROVED** [review]
Nice work! The extra diagnostic code should help with debugging in the future.  I previously missed that `checkInnerJoin` and `checkNonInnerJoin` cover the `APPLICABLE`check from the paper (`Figure 9: Pseudocode for applicable B/C`).  The inner join check looks slightly different though.  I don't know if this matters.  For example, in the paper, the check is `L-TES(◦) ⊆ S1 ∧ R-TES(◦) ⊆ S2` but we do something like `TES(◦) ⊆ S1 ∪ S2 ∧ TES(◦) ∩ S1 != ∅ ∧ TES(◦) ∩ S2 != ∅`.  The 2nd rule could be true in some cases where the first rule is false.  But the paper is dealing more with non-inner joins, so maybe that explains the difference.

:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-:-MvulOHFEEb4pMapbvMm:b5upmmj)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

**Review by DrewKimball — COMMENTED** [review]
The new formatting for `reorderjoins` looks great!

> The inner join check looks slightly different though.

For context, there's some explanation of that [here](https://github.com/cockroachdb/cockroach/blob/24d886cd713583fd6a4154e48bdbacad9eeb31f9/pkg/sql/opt/xform/join_order_builder.go#L164) and [here](https://github.com/cockroachdb/cockroach/blob/24d886cd713583fd6a4154e48bdbacad9eeb31f9/pkg/sql/opt/xform/join_order_builder.go#L1250-L1256). The gist of it is that we have to change things a little to handle the common case when different filters from an inner join predicate can be reordered differently. 

A way to build some intuition for why this works is to imagine the following process: all InnerJoin filters are pulled into a single Select operator above the join tree, then the inner joins are freely reordered. Finally, filter pushdown is used to redistribute the filters *independently* over the join tree. The code doesn't actually do this, but it achieves the same effect as if it had.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-:-MvwEXVl887yWLjWMNHK:b-107lu7)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

**Comment by mgartner** [comment]
I'm curious about another divergence from the paper in `checkNonInnerJoin`. [Here](https://github.com/cockroachdb/cockroach/blob/66d88657f6a8bcdd026735f00282953c5daa16bb/pkg/sql/opt/xform/join_order_builder.go#L1274) we check `TES(◦) ∩ S1 != ∅ ∧ TES(◦) ∩ S2 != ∅` to avoid creating new cross joins, which I believe is based on section 6.2 in the paper. However, the paper states that `𝑇(left(◦)) ∩ S1  != ∅ ∧ 𝑇(right(◦)) ∩ S2  != ∅` should be used.

I think the paper's conditional is less strict - it is true when any tables on the left and right side of the original edge, `𝑇(left(◦))` and `𝑇(right(◦))`, intersect with the vertex sets `S1` and `S2`, respectively. Our conditional is true only when tables in the egde's TES intersect with each vertex set, which is more strict because the TES is a subset of `𝑇(left(◦)) ∪ 𝑇(right(◦))`. Does this difference help prevent generating cross joins in more cases?

**Review by msirek — COMMENTED** [review]
Thanks for the explanation. It makes sense. I guess the additional intersection checks in CheckInnerJoin:`e.tes.intersects(s1) && e.tes.intersects(s2)` keep all the filters from getting pushed down one branch of the join plan tree, which might result in excessive cartesian product joins and likely would not be useful to add to the search space.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-:-Mvxk_e31KAlYUbeNBnC:b-q8yc7u)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

**Review by rytaft — APPROVED** [review]
:lgtm: nice work! (I haven't had a chance to delve deeply into the paper, so I'll let @drewkimball and/or @msirek comment on your last point)

> Reviewed 6 of 6 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76334#-:-MvxxQXIByhyzpCB16Nw:bkki0u0)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

**Comment by mgartner** [comment]
My last question is a curiosity I stumbled upon, and it doesn't necessarily relate to this change. I don't think this PR needs to hold up on its answer, so I'll go ahead and merge this.

Thanks for the reviews!

bors r+

**Comment by DrewKimball** [comment]
> I'm curious about another divergence from the paper in `checkNonInnerJoin`. [Here](https://github.com/cockroachdb/cockroach/blob/66d88657f6a8bcdd026735f00282953c5daa16bb/pkg/sql/opt/xform/join_order_builder.go#L1274) we check `TES(◦) ∩ S1 != ∅ ∧ TES(◦) ∩ S2 != ∅` to avoid creating new cross joins, which I believe is based on section 6.2 in the paper. However, the paper states that `𝑇(left(◦)) ∩ S1 != ∅ ∧ 𝑇(right(◦)) ∩ S2 != ∅` should be used.
> 
> I think the paper's conditional is less strict - it is true when any tables on the left and right side of the original edge, `𝑇(left(◦))` and `𝑇(right(◦))`, intersect with the vertex sets `S1` and `S2`, respectively. Our conditional is true only when tables in the egde's TES intersect with each vertex set, which is more strict because the TES is a subset of `𝑇(left(◦)) ∪ 𝑇(right(◦))`. Does this difference help prevent generating cross joins in more cases?

Sorry for the delay - I missed this the first time around. The check we use is exactly equivalent to the one in the paper for reasons I'll explain below*. I believe I made that change to make it more explicit that we are trying to prevent generating cross joins, but should have either stuck to the paper or added a comment to explain the change.

@mgartner since you're the one who had to read the code with fresh eyes - do you think it would be better to use the check as in the paper or add a comment? 

\* While it's true that the TES is a subset of the left and right tables, it will always include at least one table from each input because it includes tables referenced by the join filter. If the filter doesn't reference tables from one of the inputs, all tables from that input are added to the TES ([link](https://github.com/cockroachdb/cockroach/blob/6cdd96b38f5b67c655a5697b6bf493b2c59041c1/pkg/sql/opt/xform/join_order_builder.go#L1101-L1129)). Because of this handling of degenerate predicates and the requirement that referenced tables are part of S1 and S2, `T(input) ∩ S != ∅` is always equivalent to `TES(◦) ∩ S != ∅`.

TLDR: The paper's correctness relies on join filters referencing at least one table from each input. We ensure that this is the case by pretending that degenerate predicates reference all tables of the unreferenced side(s) while calculating the SES and TES.

**Comment by mgartner** [comment]
Thanks for following up! Your explanation makes sense (though it took a while to wrap my head around it all again). I think a comment that explicitly calls out the divergence from the paper and an explanation of why it's valid would be helpful. No need to use the exact check used in the paper.
